feat: apollo server 4 migration

arvi opened the pull request:
- Apollo Server migration from 3 to 4 (Apollo Server 3 EOL is Oct 2023) 
  - includes the `examples/apollo`, `README`, various imports related to Apollo Server 3 (AS3), `max-tokens` handling of token limit [graphql error](https://www.apollographql.com/docs/apollo-server/data/errors/#custom-errors). 
  
  - Regarding `max-tokens` - I have compared both `parser` and `source` values of AS3 and AS4 `graphql-armor` and they're the same (so logic for token limit shouldn't be affected). It's just that the handling of error in AS4 is different which cause a failing unit test for max token hence we added an`unexpectedErrorProcessingRequest` cycle to handle token limit error.
  
  - The main `graphql-armor` token limit error results to `Internal Server Error` however this PR makes sure it will emit `GRAPHQL_VALIDATION_FAILED` same as cost, depth limit for consistency purposes.
- Make both README.md same in locations it is found in repo
- Fix mispelled words

---

Process during development and testing
1. `bash ./install-dev.sh`
2. make modifications
3. `yarn run prettier && yarn run lint && yarn run test`
4. `yarn workspace example-apollo run dev --no-cache`
5.  Test in localhost:4000/graphql

---

This PR would benefit those who have already migrated their project to Apollo Server 4 and would not want to reintroduce Apollo Server 3 back to their repo when they want to use `graphql-armor`.

c3b5aw commented:
Hey @arvi,

I went through your MR and it seems like it is a breaking change for apollo v3 users.
Any chance we can make something which is backward compatible?

arviaccelo commented:
Hello @c3b5aw,

Here: https://github.com/Escape-Technologies/graphql-armor/pull/354/commits/954dbb34ee9253756966a8e83d0f0016b5321656

1. Added an example of `apollo v3` which is an exact copy of `main` branch `examples/apollo`. You may test it by running `yarn workspace example-apollo-v3 dev`.
2. Updated README.md for backwards compatibility

The major difference is with plugin definition having TS issue so package will determine the implementation based on version passed, no version passed (would treat it as the original implementation which is v3). Using `ApolloV4Armor` or passing a version `v4` in `ApolloArmor` would handle V4 implementation.

I used `any` for plugins instead of using either plugin definition of AS3 or AS4 as type as we do not want to re-introduce `apollo-server-core` to a GQL project that already migrated to AS4 which got rid of that package already.

arvi commented:
Hello @c3b5aw,

Here: https://github.com/Escape-Technologies/graphql-armor/pull/354/commits/954dbb34ee9253756966a8e83d0f0016b5321656

1. Added an example of `apollo v3` which is an exact copy of `main` branch `examples/apollo`. You may test it by running `yarn workspace example-apollo-v3 dev`.
2. Updated README.md for backwards compatibility

The major difference is with plugin definition having TS issue so package will determine the implementation based on version passed, no version passed (would treat it as the original implementation which is v3). Using `ApolloV4Armor` or passing a version `v4` in `ApolloArmor` would handle V4 implementation.

I used `any` for plugins instead of using either plugin definition of AS3 or AS4 as type as we do not want to re-introduce `apollo-server-core` to a GQL project that already migrated to AS4 which got rid of that package already.

c3b5aw commented on packages/graphql-armor/src/apollo/armor.ts:
```diff
@@ -24,26 +27,38 @@ export class ApolloArmor {
     validationRules: ValidationRule[];
     allowBatchedHttpRequests: false;
-    debug: false;
+    includeStacktraceInErrorResponses: false;
```
I think it can be moved over v4 only

`debug` remains false in v3 and `includeStacktraceInErrorResponses` only in v4

michael-watson commented on packages/graphql-armor/src/apollo/armor.ts:
```diff
@@ -24,26 +27,38 @@ export class ApolloArmor {
 
   protect(): {
-    plugins: PluginDefinition[];
+    plugins: any;
```
I think you can use `ApolloServerPlugin` from `@apollo/server` now

arvi replied:
@michael-watson Yep, that was actually my [initial implementation](https://github.com/Escape-Technologies/graphql-armor/pull/354/commits/639546eeec2fc2c3c096e3b621c9a32b6d494e41#diff-ce8dd078c3970c54e30c55f80357edb7d6445381483997fbaaf7f98856362182R1), but that is not backwards compatible with `Apollo Server 3`, and having TS issues so I had to put `any`. If it's not `any`, it will result to this in a project still using `Apollo Server 3`.

<img width="1449" alt="Screenshot 2023-03-14 at 6 17 42 AM" src="https://user-images.githubusercontent.com/653750/224849259-f581c8f8-c645-4588-b029-f99fa6502a8d.png">

c3b5aw replied:
```suggestion
    plugins: ApolloServerOptions<BaseContext>['plugins'];
```

would that works?

arviaccelo replied:
Hello @c3b5aw,

That change would cause issues to both AS3 and AS4 default examples like so:

Apollo Server 4
<img width="1010" alt="Screenshot 2023-03-14 at 7 33 30 AM" src="https://user-images.githubusercontent.com/111724678/224855013-63c9eaf0-0229-4cee-b2b5-04a2d2a1bfdf.png">

Apollo Server 3
<img width="1062" alt="Screenshot 2023-03-14 at 7 33 24 AM" src="https://user-images.githubusercontent.com/111724678/224855020-0cbee215-8032-4ffe-b632-0d7393f7da66.png">

arvi replied:
Hello @c3b5aw ,

That change to `plugins: ApolloServerOptions<BaseContext>['plugins'];` would cause issues to both AS4 and AS3 default examples like so:

Apollo Server 4
<img width="1010" alt="Screenshot 2023-03-14 at 7 33 30 AM" src="https://user-images.githubusercontent.com/653750/224855566-bc819d43-8a82-421b-8efe-820865117ebb.png">

Apollo Server 3
<img width="1062" alt="Screenshot 2023-03-14 at 7 33 24 AM" src="https://user-images.githubusercontent.com/653750/224855568-7b9ddbbb-28d6-4ae9-8752-03f9431a8819.png">

c3b5aw replied:
> 

You could get back to it.

`@apollo/server` has been pushed through renovate.

cf: #358 #359 

Do we have any other alternative ? I want to avoid any as much as possible.

c3b5aw replied:
Though we can duplicate that code until apollo v3 is deprecated

arvi replied:
@michael-watson @c3b5aw 

Upon checking the difference between the `ApolloServerPlugin`, I am not seeing how this is possible since they differ in `serverWillStart` type definition. 

I think AS4 is a major rewrite when I compare the difference between interfaces and the attribute types and unless AS3 packages are reinstalled back to armor and the plugins will have a `plugins: PluginDefinition[] | PluginAS3Definition[]` where it handles both, AS3 will not be compatible if not using `plugins: any`.

The purpose of the PR was to eliminate AS3 related packages for those that already migrated to AS4 😅 so I think either use `any` for `plugins` or revert to a separate class for AS4 e.g `ApolloV4Armor` use PluginDefinition there and use any for `ApolloArmor`.

Otherwise, please feel free to modify this PR should you find a way to make it work with AS3 without reintroducing AS3 packages.

ApolloServerPlugin - AS3 uses `service: GraphQLServiceContext,`
```js
export interface ApolloServerPlugin<
  TContext extends BaseContext = BaseContext,
> {
  serverWillStart?(
    service: GraphQLServiceContext,
  ): Promise<GraphQLServerListener | void>;

  requestDidStart?(
    requestContext: GraphQLRequestContext<TContext>,
  ): Promise<GraphQLRequestListener<TContext> | void>;
}
```

ApolloServerPlugin - AS4 uses `service: GraphQLServerContext,`
```js
export interface ApolloServerPlugin<
  in TContext extends BaseContext = BaseContext,
> {
  // Called once on server startup, after the schema has been loaded.
  serverWillStart?(
    service: GraphQLServerContext,
  ): Promise<GraphQLServerListener | void>;

  // Called once per request, before parsing or validation of the request has
  // occurred. This hook can return an object of more fine-grained hooks (see
  // `GraphQLRequestListener`) which pertain to the lifecycle of the request.
  requestDidStart?(
    requestContext: GraphQLRequestContext<TContext>,
  ): Promise<GraphQLRequestListener<TContext> | void>;

  /**
   * "Unexpected" errors do not include more common errors like validation,
   * parsing, or graphql execution errors. Rather, an unexpected error might
   * occur when a plugin hook throws unexpectedly or a bug in Apollo Server is
   * encountered. Notably, when errors like this occur, the error is masked to
   * the client.
   */
  unexpectedErrorProcessingRequest?({
    requestContext,
    error,
  }: {
    requestContext: GraphQLRequestContext<TContext>;
    error: Error;
  }): Promise<void>;
  // Called specifically when the user-provided `context` function throws an
  // error.
  contextCreationDidFail?({ error }: { error: Error }): Promise<void>;
  /**
   * This hook is called any time a "Bad Request" error is thrown during request
   * execution. This includes CSRF prevention and malformed requests (e.g.
   * incorrect headers, invalid JSON body, or invalid search params for GET),
   * but does not include malformed GraphQL.
   */
  invalidRequestWasReceived?({ error }: { error: Error }): Promise<void>;
  // Called on startup fail. This can occur if the schema fails to load or if a
  // `serverWillStart` or `renderLandingPage` hook throws.
  startupDidFail?({ error }: { error: Error }): Promise<void>;
}
```

---
GraphQLServerContext vs GraphQLServiceContext
```js
export interface GraphQLServerContext {
  readonly logger: Logger;
  readonly cache: KeyValueCache<string>;

  schema: GraphQLSchema;
  apollo: ApolloConfig;
  startedInBackground: boolean;
}
```
```js
export interface GraphQLServiceContext {
  logger: Logger;
  schema: GraphQLSchema;
  /**
   * @deprecated: a not particularly stable or useful hash of the schema.
   */
  schemaHash: SchemaHash;
  apollo: ApolloConfig;
  persistedQueries?: {
    cache: KeyValueCache;
  };
  serverlessFramework: boolean;
}
```

c3b5aw replied:
Alright @arvi,

Thanks a lot for your input.
Today, I merged the new apollo server packages, I'll go back to `ApolloServerPlugin`.

arvi commented on packages/graphql-armor/src/apollo/armor.ts:
```diff
@@ -48,3 +56,9 @@ export class ApolloArmor {
+  constructor(config: GraphQLArmorConfig = {}, version = 'v4') {
+    super(config, version);
+  }
+}
```
Hello @c3b5aw ,

You're right. Updated the context from `<{}>` to `<BaseContext>` and totally ditched the `ApolloServerPlugin` type and use `any` so we can just use `ApolloArmor` for both Apollo Server 3 and 4 projects instead of having an `ApolloV4Armor`.

Here: https://github.com/Escape-Technologies/graphql-armor/pull/354/commits/92218f2f5f17f1932ccfcdfcccb27914a3a3bfc2

c3b5aw commented:
Great job @arvi,

looking after the `plugins` typing, otherwise it should be ready to merge.

I'll ask one of my team member to review as we cannot afford a breaking change.

c3b5aw commented on packages/graphql-armor/package.json:
```diff
@@ -31,6 +31,7 @@
   },
   "homepage": "https://github.com/Escape-Technologies/graphql-armor",
   "dependencies": {
+    "@apollo/server": "^4.4.1",
```
I'd recommend you `pin` the dependency in `devDependencies` and lower the scope to `^4.0.0` in `optionalDependencies`

arvi replied:
Done here https://github.com/Escape-Technologies/graphql-armor/pull/354/commits/b55b40cb939e03bda81ab9bad915964b6923558a

arvi commented:
@c3b5aw Woah thanks for the merge. I was just surprised this was treated a minor change 😱 . 

This recent [commit](https://github.com/Escape-Technologies/graphql-armor/pull/354/commits/954c4644fcdb35d68cdcdce6e89fca0f9a02bb28) seems to break those with AS3 if they ever bump their version like so (as you can see it works with AS4 but not with AS3 example):

<img width="1032" alt="Screenshot 2023-03-27 at 7 50 42 AM" src="https://user-images.githubusercontent.com/653750/227812791-fb5089e2-cd6e-450b-a607-2953d9f5ab61.png">

Perhaps add a disclaimer or make this a major change for those using `dependabot` or `renovate`.
